### DISCUSSION TITLE
Multiple aquisitions in a single catalog

### DISCUSSION THREAD

**Discussion by RCBlackburn** [r-lidar/lidR#508] [Q&A]
Hi all,

I am currently attempting to run a function on a catalog composed of two different lidar acquisitions read in using:

`ctg <- readLAScatalog(c("path/to/aquistions1", "path/to/aquistions2))  `

When I run catalog_apply() with a custom function to create a chm, I end up with the following error:

```
error occurred when processing the chunk 13. Try to load this chunk with: 
chunk <- readRDS(“/tmp/RtmpDEOBMu/chunk13.rds”)
las <- readLAS(chunk)
Different files have different Point Data Format ID and are incompatible. 
```

This chunk is where the two acquisitions overlap. These acquisitions cover multiple national forests, and I would like to find a way to get around this issue without rewriting the tiles because of the large file size. 

Thanks!
Ryan

**Comment by Jean-Romain** [comment]
There is no way to by-pass this error currently. If you have las format 0 (no gpstime, no RGB) with las format 3 (gps time, RGB) they are incompatible because they do not have the same number of attributes and consequently cannot be read as one.

You may argue that if we do something like `readLAS(c("pdf0.las", "pdf3.las", select = "xyzrni")` they become compatible. Yes, certainly but I never took times to code a system that checks the compatibility of every possible pdf combination. 

Another option is to give precedence to the first file. If the first file is prf 0 and the second is pdf 3 then the gstime and RGB are simply not loaded from the second file. This actually works by default in `rlas` and the only thing that prevent doing that is a line of code in `lidR` before to start reading the files that compare the pdfs and that throws the error you have seen.  If the first file is pdf 3 and the second pdf 0, the gpstime and RGB in the second file must be set to zero because they do not exist. I'm pretty sure it also works by default in `rlas` but never tested, so I put the error you have seen.

So it may be easy to unlock a compatibility mode [here](https://github.com/r-lidar/lidR/blob/master/R/io_readLAS.R#L243) by changing

```r
if (temp.header[["Point Data Format ID"]] != header[["Point Data Format ID"]] )
   stop("Different files have different Point Data Format ID and are incompatible.", call. = FALSE)
```

to

```r
if (temp.header[["Point Data Format ID"]] != header[["Point Data Format ID"]] )
   warning("Different files have different Point Data Format ID and are incompatible. Some attribute may have been discarded or zeroed", call. = FALSE)
```

But first we need to test the behavior of 

```r
rlas::read.las(c("pdfx.las", "pdfy.las"))
# and
rlas::read.las(c("pdfy.las", "pdfx.las"))
```

Can you do that?

**Reply by RCBlackburn** [reply]
This makes sense! Thank you for the reply. I believe what you asked me to do is test the behavior of reading in two las files through the rlas from two separate catalogs (e.g. rlas::read.las(c("catalog1.las", "catalog2.las"))?

I was able to try it out changing the order of which file is read in first. Each time, I receive warnings about differences in point size and point type. The output is a dataframe with the same columns that were included in the first file. The number of rows in the dataframe does not change when the order of the files is different. I am also seeing that a certain number of points are being flagged as withheld, but the number of points being flagged stays the same regardless of the las file order.

I think that means it may be possible to use the warning over the stop?

**Reply by Jean-Romain** [reply]
> The output is a dataframe with the same columns that were included in the first file. 

Can you confirm that the attributes that are missing in the second file but that are loaded because of the first file precedence are zeroed? Can you confirm that the attributes that are present in the second file but that are non loaded because of the first file precedence are missing? Your answer is not explicit on these points.

> Each time, I receive warnings about differences in point size and point type. 

Yep, it comes from LASlib and cannot do anything for that.

> I think that means it may be possible to use the warning over the stop?

Very likely. But with a warning triggered by lidR and messages (technically not a real warnings) printed by LASlib

**Reply by RCBlackburn** [reply]
Can you confirm that the attributes that are missing in the second file but that are loaded because of the first file precedence are zeroed? 

> Yes. The attributes not present in the second file but included because of first file precedence are zeroed. 

Can you confirm that the attributes that are present in the second file but that are non loaded because of the first file precedence are missing? 

> Yes. Attributes present in the second file but not in the first file are not in the output.

**Reply by Jean-Romain** [reply]
So it is done. Install the version 3.2.4 from github. It should work

**Reply by RCBlackburn** [reply]
I just tried it out, and it works! Thank you.
